lib/FrameManager.js

```diff
@@ -631,11 +631,11 @@ class Frame {
    * @return {!Promise}
    */
   waitFor(selectorOrFunctionOrTimeout, options = {}, ...args) {
-    const xPathPattern = '//';
+    const xPathPattern = /^(?:[.][/]|[/][/]?|(?:count|id|last|name|position)[(])/i;
```

Comment:
these will break behavior for css selectors with class names "count" or others.

Reply:
Here are some samples: https://regex101.com/r/OQXuZ9/1

Reply:
The pattern for the XPath functions needs to match a `(` after the function name too, so it won't clash with CSS. Besides `count {something}` would only be usable if you had a `count` tag since the string also needs to start with it and contain a parentheses after.

Reply:
Made a small modification to also handle XPath groups, like `(//h1)[1]/text()`:

- https://regex101.com/r/OQXuZ9/2.

Reply:
> The pattern for the XPath functions needs to match a ( after the function name too, so it won't clash with CSS.

Ah, indeed. What about functional CSS selectors, e.g. `:not()` and `:nth-child()`? These require parentheses, and any of them could be added in future to CSS.

I'm concerned we'll end up with unpredictable behavior in certain cases.

Reply:
You'd need to have a CSS selector that *starts with* the function name and is *followed by* `(`.

It would be a clear break of style if a selector like that would ever be added to the CSS standard.

Even if `:count()` or something similar is eventually introduced, it won't get matched.

Reply:
@alixaxel sg. Can we have some tests that illustrate all the heuristics with this detection?

Reply:
@aslushnikov Sure, I'll write some.